presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddNotNullFiltersToJoinNode.java

```diff
@@ -0,0 +1,232 @@
+import com.facebook.presto.spi.plan.PlanNodeIdAllocator;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.PlanVariableAllocator;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.AND;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IS_NULL;
+import static com.facebook.presto.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static java.util.Collections.singletonList;
+import static java.util.Objects.requireNonNull;
+
+public class AddNotNullFiltersToJoinNode
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public AddNotNullFiltersToJoinNode(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(metadata.getFunctionAndTypeManager(), idAllocator, session), plan);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final Session session;
+        private final FunctionAndTypeManager functionAndTypeManager;
+
+        public Rewriter(FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator idAllocator, Session session)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.idAllocator = idAllocator;
+            this.session = session;
+        }
+
+        private PlanNode newJoinNode(RewriteContext<Void> context, JoinNode currentJoinNode, RowExpression newFilterExpression)
+        {
+            return new JoinNode(
+                    currentJoinNode.getSourceLocation(),
+                    idAllocator.getNextId(),
+                    currentJoinNode.getType(),
+                    context.rewrite(currentJoinNode.getLeft()),
+                    context.rewrite(currentJoinNode.getRight()),
```

Comment:
Should the rewrite to the children be called before processing of the join node in vistJoin function? This does not make a difference as in this rewriter it does not produce additional output, but I still feel it's much cleaner and less error prone to do so.

Reply:
Since each JoinNode can be re-written indpendently (as you correctly point out), this 'recursive' call looks cleaner to me IMO. I could add a comment calling this out ?  
Can you elaborate more on how this ordering could be 'error prone' ?

Reply:
Make this optimization as an iterative rule will be a better solution.